### PR TITLE
#1-4Task.Minko_Stas

### PR DESCRIPTION
**Pull request by StanislavM31**
https://travis-ci.com/github/StanislavM31/js-assignments/builds/180428310

### CODE REVIEW

File: task/01-strings-tasks.js [hunk 1]
```diff
@@ -38,6 +40,7 @@ function concatenateStrings(value1, value2) {
  *   ''      => 0
  */
 function getStringLength(value) {
+    return value.length;
     throw new Error('Not implemented');
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Extra code must be removed after implementation

File: task/01-strings-tasks.js [hunk 2]
```diff
@@ -22,6 +22,8 @@
  *   '',  'bb'  => 'bb'
  */
 function concatenateStrings(value1, value2) {
+
+    return value1+value2;
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Spaces between operators must be respected

File: task/01-strings-tasks.js [hunk 3]
```diff
@@ -129,7 +141,11 @@ function repeatString(value, count) {
  *   'I like legends', 'end' => 'I like legs',
  *   'ABABAB','BA' => 'ABAB'
  */
+/* value = not;
+str = to be or not; */
 function removeFirstOccurrences(str, value) {
+
+    return str.replace(value,'');
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Must be indent after comma

File: task/01-strings-tasks.js [hunk 4]
```diff
@@ -145,8 +161,11 @@ function removeFirstOccurrences(str, value) {
  *   '<a>' => 'a'
  */
 function unbracketTag(str) {
+
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Blank lines before and after the expression are also unnecessary.

File: task/01-strings-tasks.js [hunk 5]
```diff
@@ -145,8 +161,11 @@ function removeFirstOccurrences(str, value) {
...
+    return str.replace(/[<>]/g, '')
+
     throw new Error('Not implemented');
-}
+    }
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: Function curly brace indents must be correct

File: task/01-strings-tasks.js [hunk 6]
```diff
@@ -174,6 +194,9 @@ function convertToUpperCase(str) {
  *   'info@gmail.com' => ['info@gmail.com']
  */
 function extractEmails(str) {
+    /* return ( '[123'  str.trim()']') */;
+    var array = str.split(";");
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: In this case, you can immediately return the result without creating a variable.
Comments not related to logic can also be deleted

File: task/01-strings-tasks.js [hunk 7]
```diff
@@ -238,10 +272,24 @@ function encodeToRot13(str) {
...
+    /*попробовать вэльюоф = "string"*/
+    /* var x = new String('hi') */
+    /* === проверяет разные типы */
+    /* || typeof (new String('azaza') */
+
+    if ( value && typeof(value.valueOf()) === ('string' ) ){
+
+        return true;
+    }
+    else {
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: After "return" "else" is redundant

File: task/01-strings-tasks.js [hunk 8]
```diff
@@ -238,10 +272,24 @@ function encodeToRot13(str) {
  *   isString(new String('test')) => true
  */
 function isString(value) {
+/* проверяем только равность значений.примитивов */
+    /*попробовать вэльюоф = "string"*/
+    /* var x = new String('hi') */
+    /* === проверяет разные типы */
+    /* || typeof (new String('azaza') */
+
+    if ( value && typeof(value.valueOf()) === ('string' ) ){
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: It is necessary to observe indents and remove unnecessary indents and brackets

File: task/01-strings-tasks.js [hunk 9]
```diff
@@ -267,6 +315,20 @@ function isString(value) {
  *   'K♠' => 51
  */
 function getCardId(value) {
+
+let cards = [
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: More correct would be "const" to define

File: task/01-strings-tasks.js [hunk 10]
```diff
@@ -267,6 +315,20 @@ function isString(value) {
...
+    'A♣','2♣','3♣','4♣','5♣','6♣','7♣','8♣','9♣','10♣','J♣','Q♣','K♣',
+    'A♦','2♦','3♦','4♦','5♦','6♦','7♦','8♦','9♦','10♦','J♦','Q♦','K♦',
+    'A♥','2♥','3♥','4♥','5♥','6♥','7♥','8♥','9♥','10♥','J♥','Q♥','K♥',
+    'A♠','2♠','3♠','4♠','5♠','6♠','7♠','8♠','9♠','10♠','J♠','Q♠','K♠'
+]
+
+for (let i = 0; i < cards.length; i++) {
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: A simpler solution would be to use "indexOf"

File: task/02-numbers-tasks.js [hunk 11]
```diff
@@ -38,6 +40,8 @@ function getRectangleArea(width, height) {
  *   0    => 0
  */
 function getCicleCircumference(radius) {
+
+    return 2*Math.PI*radius;
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: It is necessary to observe indents

File: task/02-numbers-tasks.js [hunk 12]
```diff
@@ -54,6 +58,14 @@ function getCicleCircumference(radius) {
  *  -3, 3  => 0
  */
 function getAverage(value1, value2) {
+
+    let c= new Number(value1)+new Number(value2);
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: To determine the average, you need to use a simpler solution

File: task/02-numbers-tasks.js [hunk 13]
```diff
@@ -181,6 +202,31 @@ function getParallelipidedDiagonal(a,b,c) {
...
+      /*   return Math.round(num*(Math.pow(10,(-pow)))+"0"); */
+      return num;
+    }
+else{
+
+         return Math.round(num*(Math.pow(10,(-(pow))))) +"0".repeat(pow);
+}
+      /*  /* toString ([система счисления от 2 до 36]) */
+    
+    /*      let str = num.toString(10);
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: It is better to delete unnecessary commented out code

File: task/02-numbers-tasks.js [hunk 14]
```diff
@@ -181,6 +202,31 @@ function getParallelipidedDiagonal(a,b,c) {
  *   1678, 3  => 2000
  */
 function roundToPowerOfTen(num, pow) {
+    if (pow==0){
+      /*   return Math.round(num*(Math.pow(10,(-pow)))+"0"); */
+      return num;
+    }
+else{
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: "else" after "return" is redundant

File: task/02-numbers-tasks.js [hunk 15]
```diff
@@ -221,6 +287,11 @@ function isPrime(n) {
  *   toNumber(new Number(42), 0) => 42
  */
 function toNumber(value, def) {
+        if (parseInt(value)){
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: Correct indentation of conditional statement blocks must be observed.
https://javascript.info/coding-style

File: task/03-date-tasks.js [hunk 16]
```diff
@@ -56,6 +58,16 @@ function parseDataFromIso8601(value) {
  *    Date(2015,1,1)    => false
  */
 function isLeapYear(date) {
+ if(date.getFullYear()%4 !=0){
+    return false;
+   }
+    else if(date.getFullYear()%100 !=0){
+      return true;
+   }
+   else if (date.getFullYear()%400 !=0){
```
Comment (KostyaKov) [inline task/03-date-tasks.js]: "else" after "return" is redundant

File: task/02-numbers-tasks.js [hunk 17]
```diff
@@ -202,6 +248,26 @@ function roundToPowerOfTen(num, pow) {
...
+        if (n%count===0) 
+         {check = false;
+          break
+         }
+         count++
+    }
+    if (check===true) return true;
+    else return false; */
+
+      for(var i = 2; i <= Math.sqrt(n); i += 1){
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: You must use "let" instead of "var" for variables

File: task/03-date-tasks.js [hunk 18]
```diff
@@ -76,6 +88,46 @@ function isLeapYear(date) {
  *    Date(2000,1,1,10,0,0),  Date(2000,1,1,15,20,10,453)   => "05:20:10.453"
  */
 function timeSpanToString(startDate, endDate) {
+   var time = endDate - startDate;
```
Comment (KostyaKov) [inline task/03-date-tasks.js]: You must use "let" instead of "var" for variables

**Review by KostyaKov — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Review by StanislavM31 — COMMENTED** [review]
Hello, @KostyaKov. maybe this is it??

**Comment by StanislavM31** [comment]
@KostyaKov can you check this, please?

File: task/01-strings-tasks.js [hunk 19]
```diff
@@ -145,8 +147,8 @@ function removeFirstOccurrences(str, value) {
  *   '<a>' => 'a'
  */
 function unbracketTag(str) {
-    throw new Error('Not implemented');
-}
+    return str.replace(/[<>]/g, '');
+    }
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: it is necessary to observe indents.
https://javascript.info/coding-style

File: task/01-strings-tasks.js [hunk 20]
```diff
@@ -201,7 +203,8 @@ function extractEmails(str) {
  *
  */
 function getRectangleString(width, height) {
-    throw new Error('Not implemented');
+    let string = '┌'+('─'.repeat(width-2))+'┐\n'+ (('│'+' '.repeat(width-2)+'│\n').repeat(height-2)) +'└'+'─'.repeat(width-2)+'┘\n';
+    return string;
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: You can return expression result without a variable

File: task/01-strings-tasks.js [hunk 21]
```diff
@@ -221,7 +224,11 @@ function getRectangleString(width, height) {
  *
  */
 function encodeToRot13(str) {
-    throw new Error('Not implemented');
+  return str.replace(/[A-Za-z]/g, function (c) {
+    return "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz".charAt(
+           "NOPQRSTUVWXYZABCDEFGHIJKLMnopqrstuvwxyzabcdefghijklm".indexOf(c)
+           );
+        } );
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: it is necessary to observe indents.

File: task/01-strings-tasks.js [hunk 22]
```diff
@@ -238,10 +245,14 @@ function encodeToRot13(str) {
  *   isString(new String('test')) => true
  */
 function isString(value) {
-    throw new Error('Not implemented');
+    if ( value && typeof(value.valueOf()) === ('string' ) ){
```
Comment (KostyaKov) [inline task/01-strings-tasks.js]: it is necessary to observe indents and remove redundant brackets near 'string'.
Also you can return expression result without "if"
https://javascript.info/coding-style

File: task/02-numbers-tasks.js [hunk 23]
```diff
@@ -22,7 +22,7 @@
  *   5, 5  => 25
  */
 function getRectangleArea(width, height) {
-    throw new Error('Not implemented');
+    return width*height;
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: it is necessary to observe indents between operators.
https://javascript.info/coding-style

File: task/02-numbers-tasks.js [hunk 24]
```diff
@@ -54,7 +54,11 @@ function getCicleCircumference(radius) {
  *  -3, 3  => 0
  */
 function getAverage(value1, value2) {
-    throw new Error('Not implemented');
+    let c = value1 + value2;
+    if (c/2 === Infinity){
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: it is necessary to observe indents near bracket

File: task/02-numbers-tasks.js [hunk 25]
```diff
@@ -181,8 +186,13 @@ function getParallelipidedDiagonal(a,b,c) {
...
  */
 function roundToPowerOfTen(num, pow) {
-    throw new Error('Not implemented');
-}
+
+    /* попробовать с двойным == */
+    if (pow === 0){
+        return num;
+    }
+        return (Math.round(num/Math.pow(10, pow))*(Math.pow(10, pow)));
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: it is necessary to observe indents

File: task/02-numbers-tasks.js [hunk 26]
```diff
@@ -202,7 +212,12 @@ function roundToPowerOfTen(num, pow) {
  *   17 => true
  */
 function isPrime(n) {
-    throw new Error('Not implemented');
+    for(let i = 2; i <= Math.sqrt(n); i += 1){
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: it is necessary to observe indents.
Example: "for (...) {"

File: task/02-numbers-tasks.js [hunk 27]
```diff
@@ -221,7 +236,10 @@ function isPrime(n) {
  *   toNumber(new Number(42), 0) => 42
  */
 function toNumber(value, def) {
-    throw new Error('Not implemented');
+    if (parseInt(value)){
```
Comment (KostyaKov) [inline task/02-numbers-tasks.js]: it is necessary to observe indents.
Example: "if (...) {"

File: task/03-date-tasks.js [hunk 28]
```diff
@@ -76,7 +85,31 @@ function isLeapYear(date) {
...
+   let min = Math.floor((time /(1000*60))%60);
+   let hours = Math.floor((time / (1000*60*60)) %24);
+
+   if(startDate.getDay()!== endDate.getDay()){
+      hours = hours+24;
+   }
+   if(hours <10){
+       hours = "0" + hours;
+   }
+   if(min<10){
```
Comment (KostyaKov) [inline task/03-date-tasks.js]: it is necessary to observe indents.
Example: "if (min < 10) {"
https://javascript.info/coding-style

File: task/04-arrays-tasks.js [hunk 29]
```diff
@@ -337,7 +384,11 @@ function sortDigitNamesByNumericOrder(arr) {
  *   [ 1, 10, 100, 1000 ]  => 1111
  */
 function getItemsSum(arr) {
-   throw new Error('Not implemented');
+   if(arr.length>0){
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: it is necessary to observe indents.
Example: "if (a > b) {"
https://javascript.info/coding-style

File: task/04-arrays-tasks.js [hunk 30]
```diff
@@ -353,7 +404,15 @@ function getItemsSum(arr) {
  *  [ null, undefined, NaN, false, 0, '' ]  => 6
  */
 function getFalsyValuesCount(arr) {
-   throw new Error('Not implemented');
+   let count = arr.length;
+   arr.filter(function(elem){
+      if (elem){
+         count--;
+         return elem;
+         }
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js:411]: it is necessary to observe indents near bracket
https://javascript.info/coding-style

File: task/04-arrays-tasks.js [hunk 31]
```diff
@@ -353,7 +404,15 @@ function getItemsSum(arr) {
  *  [ null, undefined, NaN, false, 0, '' ]  => 6
  */
 function getFalsyValuesCount(arr) {
-   throw new Error('Not implemented');
+   let count = arr.length;
+   arr.filter(function(elem){
+      if (elem){
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: it is necessary to observe indents near bracket
https://javascript.info/coding-style

File: task/04-arrays-tasks.js [hunk 32]
```diff
@@ -454,7 +536,16 @@ function getIdentityMatrix(n) {
  *     3, 3   => [ 3 ]
  */
 function getIntervalArray(start, end) {
-   throw new Error('Not implemented');
+   let newArr = [];
+   function createArrayOfNumbers(){
+      if (end>=start){
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: it is necessary to observe indents.
Example: "if (a > b) {"
https://javascript.info/coding-style

File: task/04-arrays-tasks.js [hunk 33]
```diff
@@ -559,7 +669,19 @@ function getElementByIndexes(arr, indexes) {
  * 
  */
 function swapHeadAndTail(arr) {
-    throw new Error('Not implemented');
+   const point = arr.length;
+   if(point === 1) return arr;
+   if(point%2 !== 0){
```
Comment (KostyaKov) [inline task/04-arrays-tasks.js]: it is necessary to observe indents.
https://javascript.info/coding-style

**Review by KostyaKov — COMMENTED** [review]
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Review by KostyaKov — COMMENTED** [review]
Closing the PR as accepted! Thanks!

**Pull request closed without merging**